### PR TITLE
Isolated Galaxy - Feedback

### PR DESCRIPTION
**Pull request by aemerick**
Large PR containing changes I've made to implement and isolated galaxy propblem type following that done in the Agora isolated galaxy problem type in Enzo. I'll list a high level summary of all the changes in this PR below. 

This is a heavy WIP and depends strongly on other PRs (including solver-dd).

### CODE REVIEW

**Review by jwise77 — CHANGES_REQUESTED** [review]
Overall this is great! I only had very minor nitpicking requests, and I had a few questions in the inline comments. Also,

- Does this PR contain the changes from #43 because I saw several changes to consider all particles with mass instead of just DM particles.
- I'd remove `enzo-fire2.diff` and put it somewhere online and link it here in the PR description to make a record of it.

**Comment by jwise77** [comment]
Also to the other reviewers, there are a lot of (necessary) whitespace changes, so the review goes by much faster than you'd expect.

**Comment by bwoshea** [comment]
@aemerick In the initial PR note you stated "This is a heavy WIP and depends strongly on other PRs."  Which open PRs does this depend on, and which is it redundant with?  I'm particularly curious about PR #31 ("Required fields") and PR #42  ("Color fields to refresh in PPM"), but there might be others - your guidance would be appreciated.

**Comment by bwoshea** [comment]
@fearmayo I added you as a reviewer to this PR - please look through when you get a minute!

File: src/Enzo/enzo_EnzoInitialBurkertBodenheimer.cpp
Comment (aemerick) [reply src/Enzo/enzo_EnzoInitialBurkertBodenheimer.cpp:78]: I'm not actually sure about this. I have not tried

File: src/Enzo/enzo_EnzoInitialIsolatedGalaxy.cpp
Comment (aemerick) [reply src/Enzo/enzo_EnzoInitialIsolatedGalaxy.cpp:708]: Currently this is setting the velocity of un-deposited cells = to the avg. velocity of any neighboring cells with deposited mass...... Are you suggesting changing the if statement there to: something like `if (iflag[i] && (no neighboring cells have deposited mass)) continue`? So only totally isolated, deposited cells don't get averaged?

**Comment by aemerick** [comment]
Question on the test suite. I'm getting a lot of failures that look like this: `failed in startup: test/AmrPpm/test_adapt-L5-P1.unit` (both on CircleCI and locally) and the output from these tests looks something like this:

aemerick@localhost: Permission denied (publickey,password).
Charmrun> Error 255 returned from remote shell (localhost:0)
Charmrun> Reconnection attempt 1 of 3
aemerick@localhost: Permission denied (publickey,password).
Charmrun> Error 255 returned from remote shell (localhost:0)
Charmrun> Reconnection attempt 2 of 3
aemerick@localhost: Permission denied (publickey,password).
Charmrun> Error 255 returned from remote shell (localhost:0)
Charmrun> Reconnection attempt 3 of 3
aemerick@localhost: Permission denied (publickey,password).
Charmrun> Error 255 returned from remote shell (localhost:0)
Charmrun> Too many reconnection attempts; bailing out
Charmrun> scalable start enabled.

What does this mean? 

In addition it looks like a few tests are failing with errors on the total amount of time taken to compute. Is this a concern?

**Comment by mabruzzo** [comment]
> Question on the test suite. I'm getting a lot of failures that look like this: `failed in startup: test/AmrPpm/test_adapt-L5-P1.unit` (both on CircleCI and locally) and the output from these tests looks something like this:
> 
> aemerick@localhost: Permission denied (publickey,password).
> Charmrun> Error 255 returned from remote shell (localhost:0)
> Charmrun> Reconnection attempt 1 of 3
> aemerick@localhost: Permission denied (publickey,password).
> Charmrun> Error 255 returned from remote shell (localhost:0)
> Charmrun> Reconnection attempt 2 of 3
> aemerick@localhost: Permission denied (publickey,password).
> Charmrun> Error 255 returned from remote shell (localhost:0)
> Charmrun> Reconnection attempt 3 of 3
> aemerick@localhost: Permission denied (publickey,password).
> Charmrun> Error 255 returned from remote shell (localhost:0)
> Charmrun> Too many reconnection attempts; bailing out
> Charmrun> scalable start enabled.
> 
> What does this mean?
> 
> In addition it looks like a few tests are failing with errors on the total amount of time taken to compute. Is this a concern?

So the ssh errors are related to an error that crept in during the testing overhaul. Basically `++local` are no longer being passed to charmrun on circleci in the tests where charmrun is run in parallel. I think that's going to be addressed in PR #73.

The fact that the VL+CT test timed out is weird and has never happened before. I'll try to look at the tomorrow

File: src/Enzo/enzo_EnzoInitialBurkertBodenheimer.cpp
Comment (fearmayo) [reply src/Enzo/enzo_EnzoInitialBurkertBodenheimer.cpp:78]: Definitely best just to add a fail for 1D and 2D.  The principle purpose of this test is to test star formation in 3D I don't think it really makes sense in not 3D.

**Comment by fearmayo** [comment]
> Question on the test suite. I'm getting a lot of failures that look like this: `failed in startup: test/AmrPpm/test_adapt-L5-P1.unit` (both on CircleCI and locally) and the output from these tests looks something like this:
> 
> aemerick@localhost: Permission denied (publickey,password).
> Charmrun> Error 255 returned from remote shell (localhost:0)
> Charmrun> Reconnection attempt 1 of 3
> aemerick@localhost: Permission denied (publickey,password).
> Charmrun> Error 255 returned from remote shell (localhost:0)
> Charmrun> Reconnection attempt 2 of 3
> aemerick@localhost: Permission denied (publickey,password).
> Charmrun> Error 255 returned from remote shell (localhost:0)
> Charmrun> Reconnection attempt 3 of 3
> aemerick@localhost: Permission denied (publickey,password).
> Charmrun> Error 255 returned from remote shell (localhost:0)
> Charmrun> Too many reconnection attempts; bailing out
> Charmrun> scalable start enabled.
> 
> What does this mean?
> 
> In addition it looks like a few tests are failing with errors on the total amount of time taken to compute. Is this a concern?

Yes #PR73 should address this issue. As Matthew said it's because ++local was not being passed as an argument (not entirely sure why). I've added a fix now.

**Review by fearmayo — APPROVED** [review]

**Comment by fearmayo** [comment]
I used this PR during the summer and it definitely needs to be pulled in to the mainline - lots  of good stuff in here. I approved the PR but we need to make sure the tests are passing before merging.Documentation is also light but I would suggest that making documentation (in general) an issue might make sense there.

**Comment by fearmayo** [comment]
> I used this PR during the summer and it definitely needs to be pulled in to the mainline - lots of good stuff in here. I approved the PR but we need to make sure the tests are passing before merging.Documentation is also light but I would suggest that making documentation (in general) an issue might make sense there.

Also @stefanarridge is testing the star formation algorithms at the moment so integration tests of the star formation are in the oven

File: config/linux_gnu.py
Comment (mabruzzo) [inline config/linux_gnu.py:11]: The reason that the VL+CT tests time out is that this line became un-commented. The particular test that timed out executes 9 simulations. Turning off all optimizations really made all of the simulation time add up.

I sshed into circleci to test this. The test runs in just under 2-minutes if you comment this line again.

Comment (mabruzzo) [reply config/linux_gnu.py:11]: As a minor followup, it's not currently my intention to review this PR. I'm not really familiar with the particle machinery or any of the Enzo functionallity that's being ported. 

So I'm not sure how much useful my review would actually be.

**Comment by bwoshea** [comment]
@aemerick PR #73 has now been accepted; could you merge that?  It ought to fix the testing errors!

File: src/Enzo/enzo_EnzoMethodPmDeposit.cpp
Comment (stefanarridge) [inline src/Enzo/enzo_EnzoMethodPmDeposit.cpp:180]: I find this very confusing, why is 'dens' used to refer to particle mass, but only in the case where mass is constant?

File: src/Enzo/enzo_EnzoMethodPmDeposit.cpp
Comment (stefanarridge) [inline src/Enzo/enzo_EnzoMethodPmDeposit.cpp:204]: Now pdens is used for particle mass? But is an array of 1s when the mass is constant?

File: src/Enzo/enzo_EnzoMethodPmDeposit.cpp
Comment (stefanarridge) [inline src/Enzo/enzo_EnzoMethodPmDeposit.cpp:283]: This doesn't make much sense to me. It should be something like 'density += pmass * inv_vol * x0 * y0'. The thing that needs to be checked is whether densities and volumes are comoving.

File: config/linux_gnu.py
Comment (aemerick) [reply config/linux_gnu.py:11]: Thanks for this catch!

File: src/Enzo/enzo_EnzoMethodPmDeposit.cpp
Comment (aemerick) [reply src/Enzo/enzo_EnzoMethodPmDeposit.cpp:180]: I agree... but I think this is an idiom left over from Enzo where particle masses were stored as densities (mass / volume of a single cell)

Comment (aemerick) [reply src/Enzo/enzo_EnzoMethodPmDeposit.cpp:180]: But also this routine takes the particle mass and deposits it onto the grid as a density, so not totally incorrect to have things named as dens

File: src/Enzo/enzo_EnzoMethodPmDeposit.cpp
Comment (aemerick) [reply src/Enzo/enzo_EnzoMethodPmDeposit.cpp:204]: This is effectively a hack I used to deal with both potential cases for particle mass without having to write an entire separate routine for each case.... this is when either A) the particle type has a constant mass (and therefor no mass attribute), or B) the particle type does not have a constant mass and this must be obtained from the mass attribute.... using the `dens` variable and `pdens` array, this can be done such that when particle mass is constant, `dens` contains this value and `pdens` is 1's, so is effectively ignored, and when particle mass is not constant, `pdens` holds the particle masses and `dens` is 1 so is effectively ignored

File: src/Enzo/enzo_EnzoMethodPmDeposit.cpp
Comment (aemerick) [reply src/Enzo/enzo_EnzoMethodPmDeposit.cpp:283]: Good question. This bit of the code I just mimicked from what was already there.... @jwise77 since you looked do you have insight on this (since I'm a bit tight on time I can spend diving into this)

**Comment by aemerick** [comment]
@jwise77  I think I addressed all of your comments (though see the comment I tagged you in)

File: src/Enzo/enzo_EnzoMethodPmDeposit.cpp
Comment (stefanarridge) [reply src/Enzo/enzo_EnzoMethodPmDeposit.cpp:180]: I think this is something that needs to be cleared up... it's very confusing to have an attribute called 'mass' which is really density. Perhaps a way to fix it would be to have separate 'mass' and 'density' attributes, and this function could check for the separate cases. But this is something I can work on myself.

File: src/Enzo/enzo_EnzoMethodPmDeposit.cpp
Comment (stefanarridge) [reply src/Enzo/enzo_EnzoMethodPmDeposit.cpp:204]: Okay I can see that. This is related to my other comments on particle masses vs densities. It's something that needs to be cleared up but I can try and do this myself.

File: src/Enzo/enzo_EnzoMethodPmDeposit.cpp
Comment (stefanarridge) [reply src/Enzo/enzo_EnzoMethodPmDeposit.cpp:283]: Again, this is related to my other comments. I bring this up because in a problem I was running I was creating star particles, with masses that had actual dimensions of mass, the code was crashing since the grid densities calculated here were completely wrong. The way I fixed it was to put in an 'inv_vol' term. However, I only tested it in the case of a unigrid with no cosmology. As I said in my other comments, the issue of particle masses vs densities needs to be cleared up in my opinion.

File: src/Enzo/enzo_EnzoMethodPmDeposit.cpp
Comment (bwoshea) [reply src/Enzo/enzo_EnzoMethodPmDeposit.cpp:180]: Is this perhaps something that we can note in comments around the code in question, create an issue in the issue tracker, and then move forward?  @stefanarridge , I understand your concerns but this is a long-standing syntactic problem that is an Enzo-E issue more generally; it's not necessary to get it fixed in this PR, I don't think.

**Review by stefanarridge — APPROVED** [review]

File: src/Enzo/enzo_EnzoMethodPmDeposit.cpp
Comment (jwise77) [reply src/Enzo/enzo_EnzoMethodPmDeposit.cpp:283]: In our last meeting, we agreed that this is an important point to raise for particles in Enzo-E. I don't think this should hold up accepting this PR. I've opened up issue #83 for a discussion there.

File: src/Enzo/enzo_EnzoInitialIsolatedGalaxy.cpp
Comment (jwise77) [reply src/Enzo/enzo_EnzoInitialIsolatedGalaxy.cpp:708]: Thanks for clearing that up for me. I was misreading the code. It's fine as-is.

**Review by jwise77 — APPROVED** [review]

**Comment by jwise77** [comment]
Alright, we have 3 approvals (I'm not sure why Stefan's isn't showing in the list). I'm going to merge in the main branch manually to see if this fixes the test suite failure. After it passes the suite, I think it's ready to merge it!

**Comment by stefanarridge** [comment]
> Alright, we have 3 approvals (I'm not sure why Stefan's isn't showing in the list). I'm going to merge in the main branch manually to see if this fixes the test suite failure. After it passes the suite, I think it's ready to merge it!

I think it's because I don't have write access?

**Comment by jwise77** [comment]
> > Alright, we have 3 approvals (I'm not sure why Stefan's isn't showing in the list). I'm going to merge in the main branch manually to see if this fixes the test suite failure. After it passes the suite, I think it's ready to merge it!
> 
> I think it's because I don't have write access?

That would be the reason! I've invited you to the group.

**Comment by jwise77** [comment]
I've addressed the issue where several tests were failing. This was caused by an additional mesh update() in the initial cycle. Originally this was done to move the particles to the child blocks that were read on level 0 blocks. However, this caused the block structure to change for all tests, causing the tests to fail because the answer changed. I reverted this addition and only initialized the leaf blocks with particles instead. There were a few other bugfixes that I've included, in addition to fixes to the input file and adding an input file with particles.

**Comment by jwise77** [comment]
The tests are not failing because #81 but all of the tests pass on my local machine. I'm attaching the test output here and merging!
[PR49-testing.txt](https://github.com/enzo-project/enzo-e/files/6510411/PR49-testing.txt)

**Pull request merged**
